x/bank/keeper/keeper.go

```diff
@@ -398,28 +373,40 @@ func (k BaseKeeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins)
 		panic(sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
 	}
 
-	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amt)
+	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amounts)
 	if err != nil {
 		return err
 	}
 
-	// update total supply
-	supply := k.GetTotalSupply(ctx)
-	supply = supply.Sub(amt)
-
-	k.setSupply(ctx, supply)
+	var newSupply sdk.Coins
+	for _, amount := range amounts {
+		supply := k.GetSupply(ctx, amount.GetDenom())
+		supply = supply.Sub(amount)
+		newSupply = append(newSupply, supply)
+	}
+	k.setSupply(ctx, newSupply)
 
 	logger := k.Logger(ctx)
-	logger.Info("burned tokens from module account", "amount", amt.String(), "from", moduleName)
+	logger.Info("burned tokens from module account", "amount", amounts.String(), "from", moduleName)
 
 	// emit burn event
 	ctx.EventManager().EmitEvent(
-		types.NewCoinBurnEvent(acc.GetAddress(), amt),
+		types.NewCoinBurnEvent(acc.GetAddress(), amounts),
 	)
 
 	return nil
 }
 
+func (k BaseKeeper) setSupply(ctx sdk.Context, coins sdk.Coins) {
```

Comment:
I would just have this function update the supply for a single denom at a time. Just `sdk.Coin` not Coins.

Reply:
I am lost Robert, wasn't this what you were suggesting?

Reply:
so, @jgimeno - how about moving
```
for _, amount := range amounts {
	supply := k.GetSupply(ctx, amount.GetDenom())
	supply = supply.Add(amount)
	newSupply = append(newSupply, supply)
}
```
into `setSupply`?

Reply:
but set supply is a function that you do directly, this is first increasing the supply and later it sets the supply, there is other case in burn where instead of Add you do Sub.

Reply:
Ah, OK. Sorry, I didn't notice. Thought it's the same. So let's use your original version.